// @wordpress/block-editor/build/components/block-removal-warning-modal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockRemovalWarningModal = BlockRemovalWarningModal;

var _element = require("@wordpress/element");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _store = require("../../store");

var _lockUnlock = require("../../lock-unlock");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockRemovalWarningModal({
  rules
}) {
  const {
    clientIds,
    selectPrevious,
    blockNamesForPrompt
  } = (0, _data.useSelect)(select => (0, _lockUnlock.unlock)(select(_store.store)).getRemovalPromptData());
  const {
    clearBlockRemovalPrompt,
    setBlockRemovalRules,
    privateRemoveBlocks
  } = (0, _lockUnlock.unlock)((0, _data.useDispatch)(_store.store)); // Load block removal rules, simultaneously signalling that the block
  // removal prompt is in place.

  (0, _element.useEffect)(() => {
    setBlockRemovalRules(rules);
    return () => {
      setBlockRemovalRules();
    };
  }, [rules, setBlockRemovalRules]);

  if (!blockNamesForPrompt) {
    return;
  }

  const onConfirmRemoval = () => {
    privateRemoveBlocks(clientIds, selectPrevious,
    /* force */
    true);
    clearBlockRemovalPrompt();
  };

  return (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Are you sure?'),
    onRequestClose: clearBlockRemovalPrompt
  }, blockNamesForPrompt.length === 1 ? (0, _element.createElement)("p", null, rules[blockNamesForPrompt[0]]) : (0, _element.createElement)("ul", {
    style: {
      listStyleType: 'disc',
      paddingLeft: '1rem'
    }
  }, blockNamesForPrompt.map(name => (0, _element.createElement)("li", {
    key: name
  }, rules[name]))), (0, _element.createElement)("p", null, blockNamesForPrompt.length > 1 ? (0, _i18n.__)('Removing these blocks is not advised.') : (0, _i18n.__)('Removing this block is not advised.')), (0, _element.createElement)(_components.__experimentalHStack, {
    justify: "right"
  }, (0, _element.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: clearBlockRemovalPrompt
  }, (0, _i18n.__)('Cancel')), (0, _element.createElement)(_components.Button, {
    variant: "primary",
    onClick: onConfirmRemoval
  }, (0, _i18n.__)('Delete'))));
}
