// @wordpress/block-editor/build/components/publish-date-time-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _date = require("@wordpress/date");

var _inspectorPopoverHeader = _interopRequireDefault(require("../inspector-popover-header"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function PublishDateTimePicker({
  onClose,
  onChange,
  ...additionalProps
}, ref) {
  return (0, _element.createElement)("div", {
    ref: ref,
    className: "block-editor-publish-date-time-picker"
  }, (0, _element.createElement)(_inspectorPopoverHeader.default, {
    title: (0, _i18n.__)('Publish'),
    actions: [{
      label: (0, _i18n.__)('Now'),
      onClick: () => onChange?.(null)
    }],
    onClose: onClose
  }), (0, _element.createElement)(_components.DateTimePicker, {
    startOfWeek: (0, _date.getSettings)().l10n.startOfWeek,
    onChange: onChange,
    ...additionalProps
  }));
}

var _default = (0, _element.forwardRef)(PublishDateTimePicker);

exports.default = _default;
