// @ariakit/core/cjs/__chunks/RS4LBZRZ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _5AWPMH4Scjs = require('./5AWPMH4S.cjs');


var _RCYVACJOcjs = require('./RCYVACJO.cjs');



var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/disclosure/disclosure-store.ts
function createDisclosureStore(props = {}) {
  var _a;
  const store = _5AWPMH4Scjs.mergeStore.call(void 0, 
    props.store,
    (_a = props.disclosure) == null ? void 0 : _a.omit("contentElement", "disclosureElement")
  );
  const syncState = store == null ? void 0 : store.getState();
  const open = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.open,
    syncState == null ? void 0 : syncState.open,
    props.defaultOpen,
    false
  );
  const animated = _RCYVACJOcjs.defaultValue.call(void 0, props.animated, syncState == null ? void 0 : syncState.animated, false);
  const initialState = {
    open,
    animated,
    animating: !!animated && open,
    mounted: open,
    contentElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.contentElement, null),
    disclosureElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.disclosureElement, null)
  };
  const disclosure = _5AWPMH4Scjs.createStore.call(void 0, initialState, store);
  disclosure.setup(
    () => disclosure.sync(
      (state) => {
        if (state.animated)
          return;
        disclosure.setState("animating", false);
      },
      ["animated", "animating"]
    )
  );
  disclosure.setup(
    () => disclosure.sync(
      (state, prev) => {
        if (!state.animated)
          return;
        const mounting = state === prev;
        const animating = mounting ? state.open : state.open !== prev.open;
        disclosure.setState("animating", animating);
      },
      ["open", "animated"]
    )
  );
  disclosure.setup(
    () => disclosure.sync(
      (state) => {
        disclosure.setState("mounted", state.open || state.animating);
      },
      ["open", "animating"]
    )
  );
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, disclosure), {
    setOpen: (value) => disclosure.setState("open", value),
    show: () => disclosure.setState("open", true),
    hide: () => disclosure.setState("open", false),
    toggle: () => disclosure.setState("open", (open2) => !open2),
    stopAnimation: () => disclosure.setState("animating", false),
    setContentElement: (value) => disclosure.setState("contentElement", value),
    setDisclosureElement: (value) => disclosure.setState("disclosureElement", value)
  });
}



exports.createDisclosureStore = createDisclosureStore;
