// @ariakit/core/cjs/__chunks/ZXPKRPGY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _K2LZWNNXcjs = require('./K2LZWNNX.cjs');



var _5AWPMH4Scjs = require('./5AWPMH4S.cjs');


var _RCYVACJOcjs = require('./RCYVACJO.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/popover/popover-store.ts
function createPopoverStore(_a = {}) {
  var _b = _a, {
    popover: otherPopover
  } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, [
    "popover"
  ]);
  const store = _5AWPMH4Scjs.mergeStore.call(void 0, 
    props.store,
    otherPopover == null ? void 0 : otherPopover.omit(
      "arrowElement",
      "anchorElement",
      "contentElement",
      "popoverElement",
      "disclosureElement"
    )
  );
  const syncState = store == null ? void 0 : store.getState();
  const dialog = _K2LZWNNXcjs.createDialogStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), { store }));
  const placement = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.placement,
    syncState == null ? void 0 : syncState.placement,
    "bottom"
  );
  const initialState = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, dialog.getState()), {
    placement,
    currentPlacement: placement,
    anchorElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.anchorElement, null),
    popoverElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.popoverElement, null),
    arrowElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.arrowElement, null),
    rendered: Symbol("rendered")
  });
  const popover = _5AWPMH4Scjs.createStore.call(void 0, initialState, dialog, store);
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, dialog), popover), {
    setAnchorElement: (element) => popover.setState("anchorElement", element),
    setPopoverElement: (element) => popover.setState("popoverElement", element),
    setArrowElement: (element) => popover.setState("arrowElement", element),
    render: () => popover.setState("rendered", Symbol("rendered"))
  });
}



exports.createPopoverStore = createPopoverStore;
